cpp/src/arrow/dataset/file_base.cc

```diff
@@ -413,9 +428,108 @@ Result<compute::ExecNode*> MakeWriteNode(compute::ExecPlan* plan,
   return node;
 }
 
+namespace {
+
+class TeeNode : public compute::MapNode, public compute::BackpressureControl {
```

Comment:
You don't need to extend `compute::BackpressureControl` here (and you don't need `backpressure_control_`.  That's only needed for the `ConsumingSink` to give the `SinkNodeConsumer` something they can hold onto to pause/resume.

Since you're calling pause/resume from the node itself you can just use `this` for your calls to `Pause`/`Resume`